### PR TITLE
Selective enabling of Spyre decompositions

### PR DESCRIPTION
**Pull request by bohnstingl**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [ ] bug
- [X] feature
- [ ] documentation
- [x] cleanup

#### What this PR does:

Similar to https://github.com/torch-spyre/torch-spyre/pull/493 this PR introduces a way to selectively enabled the spyre-specific decompositions for devices when needed. In this way, they are not interfering with any in-tree devices of PyTorch.

#### Which issue(s) this PR is related to:

https://github.com/torch-spyre/torch-spyre/issues/532

#### Special notes for your reviewer:

UPDATED:

This PR builds on top of https://github.com/torch-spyre/torch-spyre/pull/625 and https://github.com/torch-spyre/torch-spyre/pull/664, which both have landed in `main`. In particular, the enabling of the spyre-specific decompositions blends into the single consolidated context manager developed in https://github.com/torch-spyre/torch-spyre/pull/664. 

Note: The current consolidated context manager ensures that the spyre-specific components are merged-in without race conditions. However, it still needs to be investigated whether the approach is tread-safe, in the sense that compilation should be supported for `cpu` and `spyre` graphs in parallel. Thus, similar to https://github.com/torch-spyre/torch-spyre/issues/531, there is an issue for the decompositions, see https://github.com/torch-spyre/torch-spyre/issues/536

#### Does this PR introduce a user-facing change?

No

#### Additional note:

cc @yoheiueda @dgrove-oss @thoangtrvn

### CODE REVIEW

**Review by yoheiueda — COMMENTED** [review]
@bohnstingl I ran into a similar issue when enabling compilation of torch.arange on Spyre. See: https://github.com/torch-spyre/torch-spyre/pull/178#issuecomment-3635914573

In that case, no decomposition was involved (the default decomposition for torch.arange was actually disabled).

After #259 merged, torch.full compiles. I believe it takes the `else` branch here (because example_inputs is empty):

https://github.com/torch-spyre/torch-spyre/blob/0d16dee2c777f4461d69829c121bac269c11754f/torch_spyre/_inductor/patches.py#L41-L52

In this case, our Fake kernel for torch.full returns a Spyre tensor:

https://github.com/torch-spyre/torch-spyre/blob/0d16dee2c777f4461d69829c121bac269c11754f/torch_spyre/_inductor/customops.py#L189-L196

As a result, graph.device_types includes "spyre" here:

https://github.com/torch-spyre/torch-spyre/blob/0d16dee2c777f4461d69829c121bac269c11754f/torch_spyre/_inductor/patches.py#L55-L60

In eager mode, factory ops typically select their backend via [BackendSelect](https://blog.ezyang.com/2020/09/lets-talk-about-the-pytorch-dispatcher/) in the dispatcher. Could we leverage a similar mechanism at compile time to determine the target device for factory ops?

If we cannot reliably detect the target compile device for factory ops, we can still work around the issue by monkey‑patching (as noted at https://github.com/torch-spyre/torch-spyre/pull/493#issuecomment-3859537754). That said, I’d prefer to minimize monkey‑patching. If possible, it would be cleaner to pass the intended decomposition table explicitly here (e.g., via kwargs).

File: tests/_inductor/test_building_blocks.py [hunk 1]
```diff
@@ -168,3 +168,53 @@ def pytorch_fn(x, w, eps, d):
...
+            .contiguous(),  # [D, T] # work around on device broadcast limitation
+            bias.reshape(T, 1)
+            .expand(T, D)
+            .contiguous(),  # [D, T] # work around on device broadcast limitation
+            torch.full([D], F16_EPS, dtype=torch.float16),
+        ]
+
+        # NOTE: To work around reduction dimension restriction,
+        #       this version performs rms_norm along dim 0
+        #       The inputs and the output should be transposed on the host
```
Comment (ani300) [inline tests/_inductor/test_building_blocks.py]: shouldn't this already work in the current stack? I know there's been a lot of work on getting the "sparse" reductions working properly. Maybe @avery-blanchard can comment as she was fixing softmax for a similar case.

**Comment by yoheiueda** [comment]
@thoangtrvn 

I don’t think the Inductor CUDA backend requires this kind of explicit enable mechanism, and it feels like `torch_spyre.enable()` introduces behavior that’s inconsistent with how Inductor’s other backends are exposed to users.

I experimented with monkey‑patching `aot_autograd` inside `torch_spyre.__init__` → https://github.com/torch-spyre/torch-spyre/commit/50b889c037db9ddbe589f973dfb8cb2aa5eef925

After applying that patch, these are the Spyre‑related loaded modules that appear immediately after `import torch`:

```
torch_spyre
torch_spyre._hooks
torch_spyre._inductor
torch_spyre._inductor.constants
torch_spyre._inductor.patches
torch_spyre._inductor.preload
torch_spyre.codegen_ops
torch_spyre.constants
torch_spyre.fallbacks
```

Note that torch_spyre._C is not imported at this stage.

**Comment by bohnstingl** [comment]
@yoheiueda This would still allow to path-in the lowerings and decomposition only when needed, right? So one could compile graphA for spyre and then followed by graphB for cpu?

@thoangtrvn and I were discussing this issue offline and we identified three options:

1. We call the function synchronously, as we do currently in my branch
2. We use the monkey-patch feature similar to the above
3. We push the burden to the user, as in your example, where we explicitly expect the user to do `torch_spyre.enable()`.

We thought that the `torch_spyre.enable()` approach would be the cleanest, but yes it is inconsistent with how inductor for CUDA operates and also pushes burden to the user. However, on the flip side, for the example above
```
import torch
def test(end, device=None):
    return torch.arange(end, device=device, dtype=torch.float16)

compiled = torch.compile(test, backend="inductor")
```
There is no way of knowing whether we should use the spyre-specific lowerings or not and we would need to "speculate". The `torch_spyre.enable()` would provide a cleaner approach here.

File: torch_spyre/_inductor/patches.py [hunk 2]
```diff
@@ -38,7 +38,21 @@ def enable_spyre_context(example_inputs):
 
     # Ensure decorators run (custom ops/decomp/lowerings modules)
     import torch_spyre._inductor.customops  # noqa: F401
-    import torch_spyre._inductor.decompositions  # noqa: F401
+    # import torch_spyre._inductor.decompositions  # noqa: F401
+    # from torch_spyre._inductor.decompositions import (
+    #     spyre_decompositions,
```
Comment (thoangtrvn) [inline torch_spyre/_inductor/patches.py]: can we clarify any of this commented code should be kept or removed?

**Review by thoangtrvn — COMMENTED** [review]
I have two comments to remove the commented codes. Other than that, I think it is in good shape.

File: torch_spyre/_inductor/patches.py [hunk 3]
```diff
@@ -85,9 +99,39 @@ def enable_spyre_context(example_inputs):
     # disable mul_softmax_pattern and div_softmax_pattern for now
     joint_graph.pass_patterns.pop()
 
+    # # Merge Spyre-specific decompositions with any existing decompositions
+    # # Note: the decompositions additionally need to be merged in this way,
```
[the remark this answers is not in the document]
Comment (bohnstingl) [reply torch_spyre/_inductor/patches.py]: Sorry, I saw this too late. I cleaned all of this up already.

File: torch_spyre/_inductor/preload.py
[the remark this answers is not in the document]
Comment (bohnstingl) [reply torch_spyre/_inductor/preload.py]: With this PR preload.py will be removed.

File: tests/_inductor/test_inductor_ops.py [hunk 4]
```diff
@@ -961,7 +977,169 @@ def fn(t):
...
+        expected_graph_str = """\
+class <lambda>(torch.nn.Module):
+    def forward(self):
+        arange_1: "i64[65]" = torch.ops.aten.arange.default(65, device = device(type='cpu'), pin_memory = False)
+        return (arange_1,)
+"""
+        inductor_graph_str = normalize_gm(
+            backend.inductor_graphs[0].print_readable(print_output=False)
+        )
+        assert inductor_graph_str == expected_graph_str, "Graphs are not identical"
```
Comment (bohnstingl) [inline tests/_inductor/test_inductor_ops.py]: This testcase reveals a known limitation that the spyre-specific lowering and decompositions are merged-in on a per-graph basis, not per operations. Thus, if `spyre` and `cpu` operations are used in the same graph, the spyre-specific elements are used. Shall we notify the user about this in any way?
We could discuss whether this is something to address in a separate issue/PR?

Comment (thoangtrvn) [reply tests/_inductor/test_inductor_ops.py]: @bohnstingl : In theory, this can happen. However, do we really see this in practice that an explicit call .to('cpu') in LLM models is used so that it at some point it runs op-A in Spyre and at another point it runs op-A in CPU. What is the reason to use op-A in CPU when op-A is supported in Spyre?

**Review by thoangtrvn — APPROVED** [review]
The PR is in good shape.  As raised in #667 , there is a need to fix and avoid global patching to allow eager mode working in some ops (including torch.clamp) patched in decompositions.py. The tests to guard this in future change is discussed in https://github.com/torch-spyre/torch-spyre/pull/703

File: torch_spyre/_inductor/decompositions.py
Comment (yoheiueda) [inline torch_spyre/_inductor/decompositions.py]: I think `spyre_decompositions`, `spyre_decompositions_to_exclude`, and `fallback_ops` are expected to be disjoint. That seems like a reasonable assumption to me.

Right now, the global decomposition table is updated in this order:

1. Add/override `spyre_decompositions`
2. Remove `spyre_decompositions_to_exclude`
3. Remove `fallback_ops`

So when restoring the global table, applying the operations in the reverse order would make the logic more robust, even if these dictionaries accidentally overlap:

1. Restore `fallback_ops`
2. Restore `spyre_decompositions_to_exclude`
3. Restore `spyre_decompositions`

What do you think?

**Review by dgrove-oss — APPROVED** [review]
Approved.  Let's go ahead and get this merged.
